### PR TITLE
feat: Vertex skinning

### PR DESCRIPTION
**Pull request by Rhuagh**
This change is [Reviewable](https://reviewable.io/reviews/amethyst/amethyst/545)

### CODE REVIEW

**Comment by omni-viral** [comment]
@Rhuagh maybe you want to try to make it with new renderer instead?
It's almost ready.

**Comment by Rhuagh** [comment]
Tbh, most of the code is independent of the renderer, it's just a vertex shader, a single component, and some wiring. I will fix that when the new renderer lands.

**Comment by Rhuagh** [comment]
I will squash after reviews are done.

**Comment by Rhuagh** [comment]
The model is from the official gltf sample models and is free to use for testing and educational purposes. 

The model jumps to a new position because that's how the animation is defined  in the gltf. I tested it with 2 separate gltf viewers and they both gave the same results. 

There's a problem here though which I forgot about, the only pass that works atm is the flat one because for some reason the gfx version we use have a 4 vertex attribute limit. I will look into that tonight.

File: amethyst_input/src/button.rs [hunk 1]
```diff
@@ -8,14 +8,18 @@ use super::local_virtual_key_code::LocalVirtualKeyCode;
 pub enum Button {
     /// Virtual Keyboard keys, use this when the letter on the key matters
     /// more than the position of the key.
-    Key(#[serde(with = "LocalVirtualKeyCode")] VirtualKeyCode),
+    Key(#[serde(with = "LocalVirtualKeyCode")]
+    VirtualKeyCode),
```
[the remark this answers is not in the document]
Comment (Rhuagh) [reply amethyst_input/src/button.rs]: Yeah, I didn't understand why it did that either

**Comment by Rhuagh** [comment]
The last issue is fixed in https://github.com/gfx-rs/gfx/pull/1750.

File: amethyst_gltf/src/format/mod.rs [hunk 2]
```diff
@@ -497,14 +510,48 @@ fn load_node(
...
+        .unwrap_or(vec![Matrix4::identity().into(); joints.len()]);
+
+    Ok(GltfSkin {
+        joints,
+        skeleton,
+        inverse_bind_matrices,
+    })
+}
+
+fn _flip_check(uv: [f32; 2], flip_v: bool) -> [f32; 2] {
```
[the remark this answers is not in the document]
Comment (Rhuagh) [reply amethyst_gltf/src/format/mod.rs]: Yeah, sorry about that, I'm currently working mostly with python at work, and there you need to put _ at the start of functions for them to be private.. I forgot that here I was working with a real language.

File: amethyst_renderer/src/skinning.rs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    Option<Vec<Separate<JointIds>>>,
+    Option<Vec<Separate<JointWeights>>>,
+);
+
+/// Build Mesh with vertex buffer combination
+fn build_mesh_with_combo(
+    combo: AnimatedVertexBufferCombination,
+    renderer: &mut Renderer,
+) -> ::error::Result<Mesh> {
+    build_mesh_with_some!(
```
Comment (omni-viral) [inline amethyst_renderer/src/skinning.rs]: I always glad to see code can be reused 😄

File: amethyst_renderer/src/pass/pbm/separate.rs [hunk 4]
```diff
@@ -44,35 +54,74 @@ impl<'a> PassData<'a> for DrawPbmSeparate {
...
-                0,
-            )
+        let mut builder = if self.skinning {
+            effect.simple(VERT_SKIN_SRC, FRAG_SRC)
+        } else {
+            effect.simple(VERT_SRC, FRAG_SRC)
+        };
+        if self.skinning {
+            builder
+                .with_raw_vertex_buffer(
```
Comment (omni-viral) [inline amethyst_renderer/src/pass/pbm/separate.rs]: I should check that such boilerplate are taken care of in new renderer.

**Comment by Rhuagh** [comment]
Need to wait for gfx to be released with the fix before this can be pushed through.

**Comment by torkleyy** [comment]
This is a great PR, I just marked some nits. I reviewed all the files, but I didn't check out the repo and run the examples.

---

Reviewed 22 of 40 files at r1, 3 of 9 files at r2, 3 of 4 files at r3, 18 of 18 files at r4.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[amethyst_animation/src/skinning/systems.rs, line 34 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3Nf5K3xyUcPwMOgUOB:-L3Nf5K3xyUcPwMOgUOC:b673o3n) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_animation/src/skinning/systems.rs#L34)):*
> ```Rust
>             .join()
>             .map(|(joint, _)| joint.skin)
>             .collect::<HashSet<Entity>>();
> ```

I don't think this `collect` is necessary.

---

*[amethyst_gltf/src/format/mod.rs, line 147 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3NfjUo9OeeLjYNImEK:-L3NfjUo9OeeLjYNImEL:bmtgrkf) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_gltf/src/format/mod.rs#L147)):*
> ```Rust
>     // TODO: morph targets, cameras
>     // TODO: KHR_materials_common extension
>     debug!("Loading nodes");
> ```

Could include `name` here.

---

*[amethyst_renderer/src/lib.rs, line 95 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3NhKhF7aWVzYwxe5U0:-L3NhKhF7aWVzYwxe5U1:bch128b) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_renderer/src/lib.rs#L95)):*
> ```Rust
> mod config;
> #[macro_use]
> mod macros;
> ```

Local macros should go before the other modules, because that's necessary for using them. Once you want to use a macro from e.g. `cam` or `color`, you'll just get a confusing error. Thus, I think putting it above all the other modules is a good practice to avoid this.x

---

*[amethyst_renderer/src/skinning.rs, line 57 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3Ni-AYv--UCU-hMCtm:-L3Ni-AYv--UCU-hMCtn:b-e39wlp) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_renderer/src/skinning.rs#L57)):*
> ```Rust
>     combo: AnimatedVertexBufferCombination,
>     renderer: &mut Renderer,
> ) -> ::error::Result<Mesh> {
> ```

nit: Can use an import here.

---

*[amethyst_renderer/src/skinning.rs, line 79 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3Ni4kN54w25NnIV83x:-L3Ni4kN54w25NnIV83y:b-m9ykze) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_renderer/src/skinning.rs#L79)):*
> ```Rust
>     /// Create a new combo mesh creator with the given combo
>     pub fn new(combo: AnimatedVertexBufferCombination) -> Self {
>         Self { combo }
> ```

Meh, I think using `Self` for the construction of a struct looks confusing. But I think that's just me :)

---

*[amethyst_renderer/src/system.rs, line 57 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3NiYsaPd4Py8TlkvFR:-L3NiYsaPd4Py8TlkvFS:b-h95vhg) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_renderer/src/system.rs#L57)):*
> ```Rust
>             Ok(pipe) => Ok(Self::new(pipe, renderer)),
>             Err(err) => {
>                 error!("Failed creating pipeline: {}", err);
> ```

Is the error message better now? Maybe we can include both the `Display` and the `Debug` output.

---

*[amethyst_renderer/src/pass/shaded_util.rs, line 42 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3NisVxrKjiM4n0Acog:-L3NisVxrKjiM4n0Acoh:b-al07ar) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_renderer/src/pass/shaded_util.rs#L42)):*
> ```Rust
> }
> 
> unsafe impl Pod for DirectionalLightPod {}
> ```

I'm wondering, could `Pod` be replaced with `Copy`? Not related with this PR, so marking this as "Satisfied".

---

*[amethyst_renderer/src/pass/shaded_util.rs, line 54 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3Nj4kHR3VvKBMtMC1C:-L3Nj4kHR3VvKBMtMC1D:bv6qknf) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_renderer/src/pass/shaded_util.rs#L54)):*
> ```Rust
>         .join()
>         .filter_map(|light| {
>             if let Light::Point(ref light) = *light {
> ```

It might be good to have `point`, `directional`, .. methods which just return an `Option`. I think I saw this pattern before.

---

*[amethyst_renderer/src/pass/util.rs, line 67 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3NjfsLFLSkWEFlMLcK:-L3NjfsLFLSkWEFlMLcL:b-8rvx2i) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_renderer/src/pass/util.rs#L67)):*
> ```Rust
> }
> 
> pub(crate) fn add_textures(
> ```

Ah, this is great :)

---

*[amethyst_renderer/src/pass/flat/interleaved.rs, line 110 at r3](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L357e970sgD9TXz2SHv-r3-110:-L3NjtiW1PAkz3hNevHB:b-dddfqu) ([raw file](https://github.com/amethyst/amethyst/blob/07ac3e3f4ee0df64362f0986882079826bd8335a/amethyst_renderer/src/pass/flat/interleaved.rs#L110)):*
<details><summary><i>Previously, Xaeroxe (Jacob Kiesel) wrote…</i></summary><blockquote>

Woops, looks like this was left in accidentally.
</blockquote></details>

I think this is resolved, please mark it as "Satisfied" in that case @Xaeroxe

---

*[amethyst_renderer/src/pass/flat/separate.rs, line 25 at r3](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L357e970sgD9TXz2SHx-r3-25:-L3NkB6qSEhhwNIwKnyB:bjwa7hh) ([raw file](https://github.com/amethyst/amethyst/blob/07ac3e3f4ee0df64362f0986882079826bd8335a/amethyst_renderer/src/pass/flat/separate.rs#L25)):*
<details><summary><i>Previously, Xaeroxe (Jacob Kiesel) wrote…</i></summary><blockquote>

This is a good first step towards composable passes.  Thank you!
</blockquote></details>

:tada: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/545#-:-L3NlBPo4pONVdRwNRpq:b-dd3ika)*

**Comment by Rhuagh** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[amethyst_animation/src/skinning/systems.rs, line 34 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3Nf5K3xyUcPwMOgUOB:-L3O5klzZ9ddtZhpXhaf:b-m4qv3n) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_animation/src/skinning/systems.rs#L34)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

I don't think this `collect` is necessary.
</blockquote></details>

How would I make it only return unique entities in the for loop then ?

---

*[amethyst_gltf/src/format/mod.rs, line 147 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3NfjUo9OeeLjYNImEK:-L3O5tJfvT846j5_UGPm:b-3zxnxw) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_gltf/src/format/mod.rs#L147)):*
> name
How do you mean ? The name of the asset is included in the first debug line above. 

---

*[amethyst_renderer/src/lib.rs, line 95 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3NhKhF7aWVzYwxe5U0:-L3O68_t7-MIQwWj4j3H:b-896fix) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_renderer/src/lib.rs#L95)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

Local macros should go before the other modules, because that's necessary for using them. Once you want to use a macro from e.g. `cam` or `color`, you'll just get a confusing error. Thus, I think putting it above all the other modules is a good practice to avoid this.x
</blockquote></details>

Done.

---

*[amethyst_renderer/src/skinning.rs, line 57 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3Ni-AYv--UCU-hMCtm:-L3O6LQpqaZHPBo5OPi0:b-896fix) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_renderer/src/skinning.rs#L57)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

nit: Can use an import here.
</blockquote></details>

Done.

---

*[amethyst_renderer/src/system.rs, line 57 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3NiYsaPd4Py8TlkvFR:-L3O6OAD-eCAnWMnWriY:b1abwp5) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_renderer/src/system.rs#L57)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

Is the error message better now? Maybe we can include both the `Display` and the `Debug` output.
</blockquote></details>

Yeah, the issue is that the other error won't give me anything from the inner errors, so I needed to print it closer to the source to get the actual gfx error message :/

---

*[amethyst_renderer/src/pass/shaded_util.rs, line 54 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3Nj4kHR3VvKBMtMC1C:-L3O6eBpYaOo8q36lKAC:b6sr3fr) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_renderer/src/pass/shaded_util.rs#L54)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

It might be good to have `point`, `directional`, .. methods which just return an `Option`. I think I saw this pattern before.
</blockquote></details>

I'll defer that for later

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/545)*

**Comment by Rhuagh** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[amethyst_renderer/src/skinning.rs, line 79 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3Ni4kN54w25NnIV83x:-L3O6mNRWud1CjBwemFC:b-4v8aho) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_renderer/src/skinning.rs#L79)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

Meh, I think using `Self` for the construction of a struct looks confusing. But I think that's just me :)
</blockquote></details>

Well, possibly for the construction, for returning Self i much prefer this way, because it means I don't have to write a gazillion type parameters all the time.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/545)*

**Comment by torkleyy** [comment]
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[amethyst_gltf/src/format/mod.rs, line 147 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3NfjUo9OeeLjYNImEK:-L3Oe5xm1lL0nKJZQVCt:baktfg3) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_gltf/src/format/mod.rs#L147)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

> name
How do you mean ? The name of the asset is included in the first debug line above. 
</blockquote></details>

Ah right. I prefer to include the name or some id because with parallel code the output can easily mix. But I guess it's fine as is.

---

*[amethyst_renderer/src/skinning.rs, line 57 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3Ni-AYv--UCU-hMCtm:-L3OeU3E5RRBXdxxJJLh:bw9y76g) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_renderer/src/skinning.rs#L57)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Err..you didn't use the imported item though :)

---

*[amethyst_renderer/src/skinning.rs, line 79 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3Ni4kN54w25NnIV83x:-L3OebEmWaxUcEfKhWQE:b-4enonj) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_renderer/src/skinning.rs#L79)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

Well, possibly for the construction, for returning Self i much prefer this way, because it means I don't have to write a gazillion type parameters all the time.
</blockquote></details>

Yeah, I prefer `Self` for specifying the return type and the real name for constructing the value.

---

*[amethyst_renderer/src/system.rs, line 57 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3NiYsaPd4Py8TlkvFR:-L3Of2c-lA9xr5h1U1Sj:bgig4x3) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_renderer/src/system.rs#L57)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

Yeah, the issue is that the other error won't give me anything from the inner errors, so I needed to print it closer to the source to get the actual gfx error message :/
</blockquote></details>

OK. Maybe the ll errors are better, let's just leave it as is for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/545)*

**Comment by Rhuagh** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[amethyst_gltf/src/format/mod.rs, line 147 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3NfjUo9OeeLjYNImEK:-L3OkreDu_Q2DNHYTYY7:b-txri5l) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_gltf/src/format/mod.rs#L147)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

Ah right. I prefer to include the name or some id because with parallel code the output can easily mix. But I guess it's fine as is.
</blockquote></details>

I can do that. Is there a good way to add a prefix for the scope with log?

---

*[amethyst_renderer/src/skinning.rs, line 57 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3Ni-AYv--UCU-hMCtm:-L3OlDw_gWEc9NCmljEi:b-6rqipu) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_renderer/src/skinning.rs#L57)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

Err..you didn't use the imported item though :)
</blockquote></details>

Ah, i used it only one place...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/545)*

**Comment by Rhuagh** [comment]
Review status: 39 of 46 files reviewed at latest revision, 3 unresolved discussions.

---

*[amethyst_gltf/src/format/mod.rs, line 147 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3NfjUo9OeeLjYNImEK:-L3YGYITeub4PvEXW6fP:bmvagva) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_gltf/src/format/mod.rs#L147)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

I can do that. Is there a good way to add a prefix for the scope with log?
</blockquote></details>

Havent found a good way to add an id without having to send it down through the function hierarchy.

---

*[amethyst_renderer/src/skinning.rs, line 57 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3Ni-AYv--UCU-hMCtm:-L3YG8kY2P04jdTm8fbn:b-896fix) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_renderer/src/skinning.rs#L57)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

Ah, i used it only one place...
</blockquote></details>

Done.

---

*[amethyst_renderer/src/skinning.rs, line 79 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3Ni4kN54w25NnIV83x:-L3YG9YBuMwEG2e363IK:b-896fix) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_renderer/src/skinning.rs#L79)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

Yeah, I prefer `Self` for specifying the return type and the real name for constructing the value.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/545)*

**Comment by Rhuagh** [comment]
This is ready, gfx-core has been released.

**Comment by torkleyy** [comment]
:lgtm:

---

Reviewed 9 of 9 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[amethyst_renderer/src/skinning.rs, line 79 at r4](https://reviewable.io:443/reviews/amethyst/amethyst/545#-L3Ni4kN54w25NnIV83x:-L3hYCgb4s2jRu8WjXMi:b8h7fx) ([raw file](https://github.com/amethyst/amethyst/blob/64be72a22339482830945d97d21ca06122db7b9f/amethyst_renderer/src/skinning.rs#L79)):*
<details><summary><i>Previously, Rhuagh (Simon Rönnberg) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

OK, thank you!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/545#-:-L3hYHnQsEb9AOnElidi:bnfp4nl)*

**Comment by omni-viral** [comment]
I have only one concern. The size of this PR. Is it really necessary to add 5k more loc to do skinning?

**Comment by Xaeroxe** [comment]
"make the PR smaller" isn't really an actionable request.  Can you point to specific sections you feel are unnecessary?

**Comment by omni-viral** [comment]
@Xaeroxe the specific point was `gltf` file which isn't a code. My statement was stupid enough to delete it. But you somehow managed to read it although it existed less than a minute 😄

**Comment by Xaeroxe** [comment]
:smile: the Roxe is always watching!

I got an email about it and happened to be looking at my phone haha

**Pull request merged**